Modify Ammo on impact

FS-21 opened the pull request:
- Warheads can now modify ammo of the affected objects.

In `rulesmd.ini`:
```ini
[SOMEWARHEAD]   ; Warhead
Ammo=0  ; integer
```

FS-21 committed:
Initial commit

- Warheads can now modify ammo of the affected objects.

In rulesmd.ini:

[SOMEWARHEAD]   ; Warhead
AmmoModifier=0  ; integer

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

FS-21 committed:
Add missing code

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/wh-modify-victim-ammo

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

Starkku reviewed (changes requested):
Looks okay except for couple of small things.

FS-21 committed:
Applied feedback

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

FS-21 commented:
> ```ini
> AmmoModifier=0 
> ```

What do you mean, reuse the tag name for the warhead tag? I don't remember it Ammo was used somewhere

mevitar commented:
So does this mean that TransferTypes, and other attempts to make warheads modify ammo, were truly abandoned?
Not because i'm agaisnt this, just want to make sure that someone won't revive those older attempts and redo what this does.

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

CrimRecya commented on src/Ext/WarheadType/Detonate.cpp:
```diff
@@ -650,3 +654,12 @@ double WarheadTypeExt::ExtData::GetCritChance(TechnoClass* pFirer) const
 	return critChance + extraChance;
 }
+
+void WarheadTypeExt::ExtData::ApplyAmmoModifier(TechnoClass* pTarget, HouseClass* pInvokerHouse, BulletExt::ExtData* pBulletExt)
```
`BulletExt::ExtData* pBulletExt` doesn't seem to have been used.

FS-21 committed:
Applied feedback

Made requested changes:
- Removed obsolete code not used anymore.
- Renamed tag from AmmoModifier to Ammo

FS-21 committed:
Updated adoc

FS-21 commented:
Applied feedback and renamed the tag from AmmoModifier to Ammo.

FS-21 committed:
Merge remote-tracking branch 'origin/develop' into feature/wh-modify-victim-ammo

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

FS-21 committed:
Merge branch 'develop' into feature/wh-modify-victim-ammo

DeathFishAtEase commented:
I saw it mentioned above as "working fine", but I'm not sure whether it means the game doesn't crash or the feature works properly. According to my investigation, since the very first commit, this statement has had an issue of being read too late, so it should always be treated as 0 regardless of what value is set in the INI, and actual testing also yielded the expected result (no effect). Moreover, as an old PR that has not yet been reviewed and approved, it obviously needs to be retested after tidying up, so I changed its test status back to "Needs testing".

mevitar commented:
> I saw it mentioned above as "working fine", but I'm not sure whether it means the game doesn't crash or the feature works properly. According to my investigation, since the very first commit, this statement has had an issue of being read too late, so it should always be treated as 0 regardless of what value is set in the INI, and actual testing also yielded the expected result (no effect). Moreover, as an old PR that has not yet been reviewed and approved, it obviously needs to be retested after tidying up, so I changed its test status back to "Needs testing".

It means it worked properly at the moment i was testing it. Ammunition was being restored as it should have when warhead hit the targets.

If you tested it with an early build then there is no way it would have worked, because the tag name was different back then.

mevitar commented:
Also i remember @FS-21 saying he's planning to rewrite this logic, Is that still the case, or is this going to be merged anyway?

Still, i tested the newest build, works fine. Didn't try multiplayer.

DeathFishAtEase commented:
I don't intend to push it forward, especially since it currently has a "Needs discussion" label that I don't quite understand. It's just that someone else wanted to use the functionality of this PR but it didn't work as expected after they merged it themselves, and the documentation description was too brief, so I did a simple maintenance. Since FS-21 has plans to rewrite it, I think we should wait for him to push it forward.

FS-21 commented:
Sorry, but I don't remember mentioning I wanted to rewrite this small feature. In all case what I wanted is to to update all  my PRs but due to my little free time I don't know if it is tomorrow or in a week or in a month or in an hour :-S
Last thing I did it was renaming the name due to NetsuNegi feedback.

DeathFishAtEase commented:
It sounds like this PR has reached a completed state. :P
@Coronia What is this `Needs Discussion` for? I think I don't understand what discussion is already or will be taking place here.

Coronia commented:
It's still about the usual 'transfer type' thing. As @Metadorius suggested all these value tweak abilities should be made into a generic pattern that can support value change based on the victim's stat, which is why some similar PRs got rejected before

IMO we don't have to implement a whole system right now, but perhaps it's better to decide tag pattern first so it's extendable and can be re-implemented without changing the protocol once a transfer type *really* happens

Metadorius commented:
I don't know, TransferType is a cursed feature, there were like 4 attempts to do it and they all flopped.

TaranDahl commented:
Given the upcoming OpenYR, I think Phobos is nearly at the end of its lifecycle. I don't think there is much benefit in considering extensibility here right now; we might as well just get this merged quickly.

Coronia commented:
so I guess this could be merged first. And if we want to extend it to use the victim's value in the future, we could just add another tag for that like `Ammo.FromVictim` etc. Same could also apply to other properties that we want them to transfer, like `TransactionMoney`. In fact this might be a better way since thse properties are actually quite different from each other, so it's too much cost but too little gain to make them all in one grand TransferType
